tla/counterfeit-cycles/counterfeit_cycles.tla

```diff
@@ -0,0 +1,293 @@
+      \* Remove message from subnet messages.
+      /\ subnetMsgs' = Tail(subnetMsgs)
+      /\ UNCHANGED<<ledger>>
+
+LedgerReceiveMessage ==
+  \* If a message is there and the balance checks out, then update it and
+  \* send approval to the subnet. If they don't check out, add sender to
+  \* the dishonestSubnets set and don't update balances.
+  LET
+    msg == Head(ledger.msgs)
+  IN
+   /\ Len(ledger.msgs) > 0
+   
+   /\ IF msg.amount > ledger.balances[msg.from] THEN
+        
+        
+        /\ ledger' = [ledger EXCEPT
+               \* Remove msg from queue.
+               !["msgs"] = Tail(@),
+               
+               \* Subnet spent more than it has. Mark it as a bad subnet.
+               !["dishonestSubnets"] = @ \union {msg.from}
+           ]
+        /\ UNCHANGED<<subnets, subnetMsgs>>
+      ELSE
+        /\ ledger' = [
+            ledger EXCEPT
+              \* Transaction is valid. Update ledger balances.
+              !["balances"][msg.from] = @ - msg.amount,
+              !["balances"][msg.to] = @ + msg.amount,
+              
+              \* Remove msg from queue.
+              !["msgs"] = Tail(@)
+            ]
+        /\ subnetMsgs'= Append(subnetMsgs,
+            [type |-> "approve", from |-> msg.from, to |-> msg.to, amount |-> msg.amount]
+           )
+        /\ UNCHANGED<<subnets>>
+
+Next ==
+  \/ SubnetSendTransfer
+  \/ SubnetReceiveTransfer
+  \/ SubnetReceiveApprove
+  \/ SubnetDishonestSendTransfer
+  \/ LedgerReceiveMessage
+  \/ LedgerPoll
+
+BalancesNonNegative ==
+  (*************************************************************************)
+  (* Invariant to ensure that all subnet balances are non-negative.        *)
+  (*************************************************************************)
+  \A s \in SUBNETS: subnets[s].balance >= 0
+
+
+LedgerIdentifiesAllDishonestSubnets ==
+  {s \in SUBNETS: ~subnets[s].honest} = ledger.dishonestSubnets
+
+LedgerIdentifiesAllDishonestSubnetsEventually ==
+  \* A temporal property ensuring that all dishonest subnets are,
+  \* eventually, identified by the ledger.
+  \*
+  \* NOTE: I think this invariant is not relevant anymore and can be removed.
+  <>[][LedgerIdentifiesAllDishonestSubnets]_<<subnets, subnetMsgs, ledger>>
+
+RECURSIVE RecordToSeq(_,_)
+RecordToSeq(f, ks) ==
+  \* Helper method to transform a record's values into a sequence.
+  IF ks = {} THEN << >>
+  ELSE << f[CHOOSE k \in ks: TRUE] >> \o RecordToSeq(f, ks \ {CHOOSE k \in ks: TRUE})
+
+RECURSIVE SumSeq(_)
+SumSeq(s) ==
+  \* Helper method to sum a sequence
+  IF Len(s) = 0 THEN 0
+  ELSE Head(s) + SumSeq(Tail(s))
+  
+RECURSIVE SumBalance(_)
+SumBalance(s) ==
+  \* Helper method to sum a sequence
+  IF Len(s) = 0 THEN 0
+  ELSE Head(s).balance + SumBalance(Tail(s))
+
+NoFakeCyclesAreCreated ==
+  \* The total supply of finalized cycles in the ledger remains constant.
+  /\ SumSeq(RecordToSeq(ledger.balances, SUBNETS)) = STARTING_BALANCE_PER_SUBNET * Cardinality(SUBNETS)
```

Comment:
Nit: I'd define a `STARTING_SUPPLY ==  STARTING_BALANCE_PER_SUBNET * Cardinality(SUBNETS)`  somewhere

Reply:
Alternatively, you could define an "action invariant" such as this:
`SumSeq(RecordToSeq(ledger.balances, SUBNETS)) < SumSeq(RecordToSeq(ledger'.balances, SUBNETS))`
Like invariants, action invariants can still be processed fairly efficiently by TLC, unlike general safety and temporal properties which can slow the analysis down significantly (basically it's a local check on each edge of the state graph, rather than a graph search).